test/testutils.jl

```diff
@@ -0,0 +1,330 @@
+            else
+                @test isequal(impute(c, tester.imp(; tester.kwargs...)), c)
+            end
+        end
+
+        @testset "Too many missing values" begin
+            # Test Context error condition
+            c = fill(missing, 5, 2)
+            settings = deepcopy(tester.kwargs)
+            settings[:context] = Context(; limit=0.1)
+            @test_throws ImputeError impute(c, tester.imp(; settings...))
+            @test_throws ImputeError tester.f(c; settings...)
+        end
+    end
+end
+
+function test_dataframe(tester::ImputorTester)
+    @testset "DataFrame" begin
+        table = DataFrame(
+            :sin => allowmissing(sin.(1.0:1.0:20.0)),
+            :cos => allowmissing(sin.(1.0:1.0:20.0)),
+        )
+
+        table.sin[[2, 3, 7, 12, 19]] .= missing
+
+        result = impute(table, tester.imp(; tester.kwargs...))
+
+        @testset "Base" begin
+            # Test that we have fewer missing values
+            @test count(ismissing, Matrix(result)) < count(ismissing, Matrix(table))
+            @test isa(result, DataFrame)
+
+            # Test that functional form behaves the same way
+            @test result == tester.f(table; tester.kwargs...)
+        end
+
+        @testset "In-place" begin
+            # Test that the in-place function return the new results and logs whether it
+            # successfully did it in-place
+            table2 = deepcopy(table)
+            table2_ = tester.f!(table2; tester.kwargs...)
+            @test table2_ == result
+            if table2 != result
+                @warn "$(tester.f!) did not mutate input data of type DataFrame"
+            end
+        end
+
+        @testset "No missing" begin
+            # Test having no missing data
+            b = DataFrame(
+                :sin => allowmissing(sin.(1.0:1.0:20.0)),
+                :cos => allowmissing(sin.(1.0:1.0:20.0)),
+            )
+            @test impute(b, tester.imp(; tester.kwargs...)) == b
+        end
+
+        @testset "All missing" begin
+            # Test having only missing data
+            c = DataFrame(
+                :sin => fill(missing, 10),
+                :cos => fill(missing, 10),
+            )
+            if tester.imp == DropObs
+                @test impute(c, tester.imp(; tester.kwargs...)) == DataFrame()
+            elseif tester.imp == DropVars
+                # https://github.com/JuliaData/Tables.jl/issues/117
+                @test_broken impute(c, tester.imp(; tester.kwargs...)) == DataFrame()
+            else
+                @test isequal(impute(c, tester.imp(; tester.kwargs...)), c)
+            end
+        end
+
+        @testset "Too many missing values" begin
+            # Test Context error condition
+            c = DataFrame(
+                :sin => fill(missing, 10),
+                :cos => fill(missing, 10),
+            )
+            settings = deepcopy(tester.kwargs)
+            settings[:context] = Context(; limit=0.1)
```

Comment:
`tester.kwargs` should be a NamedTuple, and you should overwrite context via `merge(tester.kwargs, (; context=...))`

Reply:
:+1 `merge` is definitely the better solution than having it be a `Dict`